## Analyzing Gun Deaths in the US
The dataset can be found in the github repository.
It contains information on gun deaths in the US from 2012 to 2014.
Each row in the dataset represents a single fatality. The columns contain demographic and other information about the victim. 

### 1. Reading guns.csv file and splitting headers from data

In [2]:
import csv
import datetime

# Open the file using the open() function.
f = open("guns.csv", "r")
# Use the csv.reader() function to load the opened file.
csvreader = csv.reader(f)
# Convert csvreader variable to list and assign it to data variable.
data = list(csvreader)
# Extracting first row of data, and assign it to the variable headers.
headers = data[:1]
# Remove the first row from data.
data = data[1:]
# Display headers.
print(headers)
# Display first 5 rows of data.
print(data[:5])

[['', 'year', 'month', 'intent', 'police', 'sex', 'age', 'race', 'hispanic', 'place', 'education']]
[['1', '2012', '01', 'Suicide', '0', 'M', '34', 'Asian/Pacific Islander', '100', 'Home', '4'], ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'], ['3', '2012', '01', 'Suicide', '0', 'M', '60', 'White', '100', 'Other specified', '4'], ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4'], ['5', '2012', '02', 'Suicide', '0', 'M', '31', 'White', '100', 'Other specified', '2']]


### 2. Finding how many gun deaths occur in each year

In [3]:
# Create year_counts dictionary
year_counts = dict()
for row in data:
    # Extract year column from data
    years = row[1]
    if years in year_counts:
        # if the year element is a key in year_counts dictionary, increment the value by one. 
        year_counts[years] += 1
    else:
        # if the year element is not a key in year_counts dictionary, set the value to one. 
        year_counts[years] = 1      
# Display year_counts to see how many gun deaths occur in each year.
print(year_counts)

{'2012': 33563, '2013': 33636, '2014': 33599}


### 3. Let's see if gun deaths in the US change by month and year.

In [8]:
# Create dates list
dates = []
# Create date_counts dictionary
date_counts = dict()
for row in data:
    # Extracting year and month from data, and create datetime object from them to append to dates list.
    dates.append(datetime.datetime(year=int(row[1]), month=int(row[2]), day=1))
for row in dates:
    if row in date_counts:
        # if the date element is a key in date_counts dictionary, increment the value by one. 
        date_counts[row] += 1
    else:
        # if the date element is not a key in date_counts dictionary, set the value to one.
        date_counts[row] = 1
# Display date_counts to see change gun deaths by month and year
print(date_counts)   

{datetime.datetime(2012, 1, 1, 0, 0): 2758, datetime.datetime(2012, 2, 1, 0, 0): 2357, datetime.datetime(2012, 3, 1, 0, 0): 2743, datetime.datetime(2012, 4, 1, 0, 0): 2795, datetime.datetime(2012, 5, 1, 0, 0): 2999, datetime.datetime(2012, 6, 1, 0, 0): 2826, datetime.datetime(2012, 7, 1, 0, 0): 3026, datetime.datetime(2012, 8, 1, 0, 0): 2954, datetime.datetime(2012, 9, 1, 0, 0): 2852, datetime.datetime(2012, 10, 1, 0, 0): 2733, datetime.datetime(2012, 11, 1, 0, 0): 2729, datetime.datetime(2012, 12, 1, 0, 0): 2791, datetime.datetime(2013, 1, 1, 0, 0): 2864, datetime.datetime(2013, 2, 1, 0, 0): 2375, datetime.datetime(2013, 3, 1, 0, 0): 2862, datetime.datetime(2013, 4, 1, 0, 0): 2798, datetime.datetime(2013, 5, 1, 0, 0): 2806, datetime.datetime(2013, 6, 1, 0, 0): 2920, datetime.datetime(2013, 7, 1, 0, 0): 3079, datetime.datetime(2013, 8, 1, 0, 0): 2859, datetime.datetime(2013, 9, 1, 0, 0): 2742, datetime.datetime(2013, 10, 1, 0, 0): 2808, datetime.datetime(2013, 11, 1, 0, 0): 2758, datet

### 4. How gun deaths in the US vary by gender and race.

In [10]:
# Create sex_counts dictionary
sex_counts = dict()
for row in data:
    # Extracting sex column
    sex = row[5]
    # Count up how many times each item in the sex column occurs.
    if sex in sex_counts:
         # if the sex element is a key in sex_counts dictionary, increment the value by one. 
        sex_counts[sex] +=1
    else:
        # if the sex element is not a key in sex_counts dictionary, set the value to one.
        sex_counts[sex] = 1
# Create race_counts dictionary
race_counts = dict()
for row in data:
    race = row[7]
    # Count up how many times each item in the race column occurs.
    if race in race_counts:
         # if the race element is a key in race_counts dictionary, increment the value by one. 
        race_counts[race] +=1
    else:
        # if the race element is not a key in race_counts dictionary, set the value to one.
        race_counts[race] = 1
# Display race_counts and sex_counts
print(sex_counts)
print(race_counts)

{'M': 86349, 'F': 14449}
{'Asian/Pacific Islander': 1326, 'White': 66237, 'Native American/Native Alaskan': 917, 'Black': 23296, 'Hispanic': 9022}


### 5. Reading census.csv file

In [12]:
f = open("census.csv", "r")
csvreader = csv.reader(f)
census = list(csvreader)
# Display census data
print(census)

[['Id', 'Year', 'Id', 'Sex', 'Id', 'Hispanic Origin', 'Id', 'Id2', 'Geography', 'Total', 'Race Alone - White', 'Race Alone - Hispanic', 'Race Alone - Black or African American', 'Race Alone - American Indian and Alaska Native', 'Race Alone - Asian', 'Race Alone - Native Hawaiian and Other Pacific Islander', 'Two or More Races'], ['cen42010', 'April 1, 2010 Census', 'totsex', 'Both Sexes', 'tothisp', 'Total', '0100000US', '', 'United States', '308745538', '197318956', '44618105', '40250635', '3739506', '15159516', '674625', '6984195']]


### 6. Finding rate of gun deaths per 100000 people for each racial category

In [21]:
# Manually creating a mapping dictionary that maps each key from the race_counts 
# to the population count of the race from census. 
mapping = {
    "Asian/Pacific Islander": float(census[1][14]) + float(census[1][15]),
    "Black": float(census[1][12]),
    "Native American/Native Alaskan": float(census[1][13]),
    "Hispanic": float(census[1][11]),
    "White": float(census[1][10])
}
#Create race_per_hundredk dictionary.
race_per_hundredk = dict()
    
for key, value in race_counts.items():
    if key in mapping:
        race_per_hundredk[key] = (float(value) / float(mapping[key])) * 100000
# Display race_per_hundredk. 
# It contains the rate of gun deaths per 100000 people for each racial category.
print(race_per_hundredk)

{'Asian/Pacific Islander': 8.374309664161762, 'White': 33.56849303419181, 'Native American/Native Alaskan': 24.521955573811088, 'Black': 57.8773477735196, 'Hispanic': 20.220491210910907}


## Explanation of 6th cell:

  In order to get from the raw counts of gun deaths by race to a rate of gun deaths per 100,000 people in each race, we'll need to divide the total number of gun deaths by the population of each race. From the census dataset, we know that the number of people in the White racial category is 197318956. We'd divide 66237 by 197318956

 white_gun_death_rate = 66237 / 197318956

This gives us the percentage chance that a given person in the White census race category would have been killed by a gun in the US from 2012 to 2014. If you do this computation, you'll see that the rate is a very small number, 0.0003356849303419181. It's for this reason that it's typical to express crime statistics as the "rate per 100000". This tells you the number of people in a given group out of every 100000 that were killed by guns in the US. To get this, we just multiply by 100000.

rate_per_hundredk = 0.0003356849303419181 * 100000

This gives us 33.56, which we can interpret as "33.56 out of every 100000 people in the White census race category in the US were killed by guns between 2012 and 2014.

### 7. Finding number of gun deaths by homicide for that race

In [22]:
# Create intents list
intents = []
# Create races list
races = []
for row in data:
    # Extract the intent column
    intents.append(row[3])
    # Extract the race column
    races.append(row[7])
    
# Create homicide_race_counts dictionary
homicide_race_counts = dict()    
for i, race in enumerate(races):
    # if the intent is Homicide
    if intents[i] == "Homicide":
        if race in homicide_race_counts:
            # if the race element is a key in homicide_race_counts dictionary, increment the value by one. 
            homicide_race_counts[race] +=1
        else:
            # if the race element is not a key in homicide_race_counts dictionary, set the value to one.
            homicide_race_counts[race] = 1
                
print(homicide_race_counts)  

{'White': 9147, 'Asian/Pacific Islander': 559, 'Black': 19510, 'Native American/Native Alaskan': 326, 'Hispanic': 5634}


### 8. Finding the rate of gun deaths per 100,000 people for each race homicide category

In [23]:
#Create race_homicide_per_hundredk dictionary.
race_homicide_per_hundredk = dict()
for key, value in homicide_race_counts.items():
    if key in mapping:
        race_homicide_per_hundredk[key] = (float(value) / float(mapping[key])) * 100000
# Display race_homicide_per_hundredk. 
# It contains the rate of gun deaths per 100,000 people for each race homicide category.
print(race_homicide_per_hundredk)

{'White': 4.6356417981453335, 'Asian/Pacific Islander': 3.530346230970155, 'Black': 48.471284987180944, 'Native American/Native Alaskan': 8.717729026240365, 'Hispanic': 12.627161104219914}
